### PR TITLE
BL-1186 Rewrite of Overflow code

### PR DESCRIPTION
**Pull request by gmartin7**
* Now detects better if typing in a box pushes another box
  out of the margin box
* Works better with the origami resizing bars too

### CODE REVIEW

**Comment by gmartin7** [comment]
N.B.: StyleEditor.js and the Overflow*.htm files are generated. We just need to review the .ts and .jade versions.

**Comment by hatton** [comment]
Gordon, would you be willing to extract this stuff out to its own js file? We're trying to  reduce the bloomEditing.js monstrosity, and in lieu of one big overhaul, we've been doing extract each time we do major work. A "rewrite" of the overflow would be the best time to do that, as it's all in your head.

**Comment by gmartin7** [comment]
OK. So nobody merge this until I push a new version.

**Comment by gmartin7** [comment]
OverFlowChecker.ts is now factored out of bloomEditing.js :)

**Comment by sillsdevgerrit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.lsdev.sil.org//job/GitHub-Bloom-Wrapper-master-debug/425/
Test PASSed.

**Comment by sillsdevgerrit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.lsdev.sil.org//job/GitHub-Bloom-Wrapper-master-debug/426/
Test PASSed.

File: src/BloomBrowserUI/bookEdit/OverflowChecker/OverflowChecker.ts [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+        //Add the handler so that when the elements change, we test for overflow
+        editablePageElements.on("keyup paste", function (e) {
+            var target = e.target;
+            // Give the browser time to get the pasted text into the DOM first, before testing for overflow
+            // GJM -- One place I read suggested that 0ms would work, it just needs to delay one 'cycle'.
+            //        At first I was concerned that this might slow typing, but it doesn't seem to.
+            setTimeout(function () {
+                OverflowChecker.MarkOverflowInternal(target);
+
+                //REVIEW: why is this here, in the overflow detection?
```
Comment (JohnThomson) [inline src/BloomBrowserUI/bookEdit/OverflowChecker/OverflowChecker.ts]: Since this is new code, I would hope you know! Are you saying this was something left over from an earlier version?

File: src/BloomBrowserUI/bookEdit/OverflowChecker/OverflowChecker.ts [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+
+                // This will make sure that any language tags on this div stay in position with editing.
+                // Reposition all language tips, not just the tip for this item because sometimes the edit moves other controls.
+                (<qtipInterface>$(queryElementsThatCanOverflow)).qtip('reposition');
+            }, 100); // 100 milliseconds
```
Comment (JohnThomson) [inline src/BloomBrowserUI/bookEdit/OverflowChecker/OverflowChecker.ts]: Until the qtip moves? Until the qtip shows on hover?

File: src/BloomBrowserUI/bookEdit/OverflowChecker/OverflowChecker.ts [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+            var fontSizeRemnant = new StyleEditor("/bloom/bookEdit").GetCalculatedFontSizeInPoints(element) - 22;
+            if (fontSizeRemnant > 0) {
+                growFromCenterVerticalFudgeFactor = (fontSizeRemnant / 5) + 1;
+            }
+        }
+
+        //In the Picture Dictionary template, all words have a scrollHeight that is 3 greater than the client height.
+        //In the Headers of the Term Intro of the SHRP C1 P3 Pupil's book, scrollHeight = clientHeight + 6!!! Sigh.
+        // the focussedBorderFudgeFactor takes care of 2 pixels, this adds one more.
+        var shortBoxFudgeFactor = 4;
```
Comment (JohnThomson) [inline src/BloomBrowserUI/bookEdit/OverflowChecker/OverflowChecker.ts]: This fudge factor appears to be being applied everywhere, not just in picture dictionaries.

File: src/BloomBrowserUI/bookEdit/OverflowChecker/OverflowChecker.ts [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+            });
+        });
+
+        // Right now, test to see if any are already overflowing
+        editablePageElements.each(function () {
+            OverflowChecker.MarkOverflowInternal(this);
+        });
+    }
+
+    // Actual testable determination of Type I overflow or not
```
Comment (JohnThomson) [inline src/BloomBrowserUI/bookEdit/OverflowChecker/OverflowChecker.ts]: If you explained somewhere what Type 1 and 2 overflow are, I missed it. An explanation or cross-ref would be helpful.

File: src/BloomBrowserUI/bookEdit/OverflowChecker/OverflowChecker.ts [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+        var $box = $(box);
+        $box.removeClass('overflow');
+        if (OverflowChecker.IsOverflowingSelf(box)) {
+            $box.addClass('overflow');
+        } else {
+            //now, thing is, while the text may fit in our box, our box may not fit our parent. Or grandparent, etc.
+            //It could be that we could just do this on up the hierarchy? For now, here's the case we know is important,
+            // in the Origami pages, where the space allocated could be too small.
+            if (OverflowChecker.HasSplitParentThatOverflows($box)) {
+                $box.addClass('overflow');
```
Comment (JohnThomson) [inline src/BloomBrowserUI/bookEdit/OverflowChecker/OverflowChecker.ts]: Since you do the same thing in both 'if' bodies, this feels more like a 'this or that' condition than nested ifs. (i.e., you could move the second condition into the outer if using ||)

File: src/BloomBrowserUI/bookEdit/OverflowChecker/OverflowChecker.ts [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+        var container = $box.closest('.marginBox');
+        //NB: for some historical reason in March 2014 the calendar still uses textareas
+        var queryElementsThatCanOverflow = ".bloom-editable, textarea";
+        var editablePageElements = $(container).find(queryElementsThatCanOverflow);
+
+        // Type 2 Overflow - We'll check ALL of these for overflow past marginBox
+        editablePageElements.each(function() {
+            var $this = $(this);
+            if (OverflowChecker.IsOverflowingMargins($this[0])) {
+                $box.addClass('overflow'); // probably typing in the focused element caused this
```
Comment (JohnThomson) [inline src/BloomBrowserUI/bookEdit/OverflowChecker/OverflowChecker.ts]: How would $box lose its overflow marker, if the user edited some other box to solve the parent overflow problem?

File: src/BloomBrowserUI/bookEdit/OverflowChecker/OverflowChecker.ts [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+                if (!OverflowChecker.IsOverflowingSelf($this[0]) && !OverflowChecker.HasSplitParentThatOverflows($this)) {
+                    $this.removeClass('overflow'); // might be a remnant from earlier overflow
+                }
+            }
+        });
+        OverflowChecker.UpdatePageOverflow(container.closest('.bloom-page'));
+    } // end MarkOverflowInternal
+
+    static HasSplitParentThatOverflows(jQueryBox : JQuery) : boolean {
+        var splitterParents = jQueryBox.parents('.split-pane-component-inner');
```
Comment (JohnThomson) [inline src/BloomBrowserUI/bookEdit/OverflowChecker/OverflowChecker.ts]: The name of this method suggests it is checking all parents, but it actually only checks the first. ParentIsSplitThatOverflows? might better suggest that it's only looking at the most immediate parent.
Is that all you need to consider?

File: src/BloomBrowserUI/bookEdit/OverflowChecker/OverflowChecker.ts (hunk 2 above)
Comment (gmartin7) [inline src/BloomBrowserUI/bookEdit/OverflowChecker/OverflowChecker.ts]: The 100ms delay is to give the browser time to get the pasted text into the DOM before running the overflow check (see the comment above setTimeout).
The REVIEW comment is from before the refactoring and refers (I believe) to the qtip repositioning. I thought about just making a separate 'paste' handler for the qtip repositioning, but... the code would all be the same and actually the setTimeout is only needed (afaik) for the paste, but the qtip reposition might be useful for both typing and pasting... so I left it as is.

File: src/BloomBrowserUI/bookEdit/OverflowChecker/OverflowChecker.ts (hunk 3 above)
Comment (gmartin7) [inline src/BloomBrowserUI/bookEdit/OverflowChecker/OverflowChecker.ts]: true. This one was added by JH some time ago, I believe. This is old refactored code, not new stuff, for the most part. Careful reading of JH's comment seems to indicate he found 1 additional  pixel in the Picture Dictionary and then 3 more in some Uganda book he had done... No wonder he wrote 'Sigh.' :)

File: src/BloomBrowserUI/bookEdit/OverflowChecker/OverflowChecker.ts (hunk 5 above)
Comment (gmartin7) [inline src/BloomBrowserUI/bookEdit/OverflowChecker/OverflowChecker.ts]: Most interesting! I wonder how I missed that. Will do.

File: src/BloomBrowserUI/bookEdit/OverflowChecker/OverflowChecker.ts (hunk 3 above)
Comment (JohnThomson) [inline src/BloomBrowserUI/bookEdit/OverflowChecker/OverflowChecker.ts]: So you're saying that it was being applied everywhere even before you pulled overflowchecker into its own file?

File: src/BloomBrowserUI/bookEdit/OverflowChecker/OverflowChecker.ts (hunk 2 above)
Comment (JohnThomson) [inline src/BloomBrowserUI/bookEdit/OverflowChecker/OverflowChecker.ts]: The point of my comment was to suggest that the 100ms delay needs some explanation in a code comment.
Surely though it has to do with positioning qtips not checking overflow?

File: src/BloomBrowserUI/bookEdit/OverflowChecker/OverflowChecker.ts (hunk 6 above)
Comment (gmartin7) [inline src/BloomBrowserUI/bookEdit/OverflowChecker/OverflowChecker.ts]: That's what the 'else' just below this is for (at least partially). If they go and fix the overflow elsewhere, then $box will fall into the category of the else when the edited box is processed for overflow and its overflow marker will be removed there.

File: src/BloomBrowserUI/bookEdit/OverflowChecker/OverflowChecker.ts (hunk 7 above)
Comment (gmartin7) [inline src/BloomBrowserUI/bookEdit/OverflowChecker/OverflowChecker.ts]: I'll add 'Immediate' to the method name and I moved Andrew's comments about origami checking down from MarkOverflowInternal.

**Comment by sillsdevgerrit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.lsdev.sil.org//job/GitHub-Bloom-Wrapper-master-debug/438/
Test PASSed.

File: src/BloomBrowserUI/bookEdit/OverflowChecker/OverflowChecker.ts (hunk 6 above)
Comment (JohnThomson) [inline src/BloomBrowserUI/bookEdit/OverflowChecker/OverflowChecker.ts]: Fair enough. For some reason I misread things and thought it was only checking boxes that had been edited, but you're hitting everything that's capable of overflowing.

File: src/BloomBrowserUI/bookEdit/OverflowChecker/OverflowChecker.ts (hunk 3 above)
Comment (gmartin7) [inline src/BloomBrowserUI/bookEdit/OverflowChecker/OverflowChecker.ts]: Yes.
